### PR TITLE
Fix anonymous records internal error when nested with overlapping labels

### PR DESCRIPTION
**Pull request by abonie**
This will fix secondary issue reported here https://github.com/dotnet/fsharp/issues/6411

`AnonRecdTypeInfo` was concatenating all of the record's sorted labels to create a key for some post inference check. If for two records concatenating their labels gave the same result, then the key was also the same which resulted in some collisions.

### CODE REVIEW

**Review by psfinaki — APPROVED** [review]
Nice one

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/Misc/AnonRecd.fs.il.net472.bsl [hunk 1]
```diff
@@ -79,13 +79,13 @@
 
 } 
 
-.class public auto ansi serializable sealed beforefieldinit '<>f__AnonymousType1912756633`2'<'<A>j__TPar','<B>j__TPar'>
+.class public auto ansi serializable sealed beforefieldinit '<>f__AnonymousType3376895005`2'<'<A>j__TPar','<B>j__TPar'>
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/Misc/AnonRecd.fs.il.net472.bsl:82]: What does this change do to the "By Default Works Across Assembly Boundaries " principle https://github.com/fsharp/fslang-design/blob/main/FSharp-4.6/FS-1030-anonymous-records.md#design-principle-by-default-works-across-assembly-boundaries of F# anonymous records?

So either:
- New code with new compiler working with a .dll that was built with the older version
- Or, older compiler working with .dll that was produced with the new one

(let's say the .dll exposes a function which expects a particular anonymous record)

**Review by T-Gro — DISMISSED** [review]
Postponing until we get a clarification/test/.. on how this affects name stability.

The RFC explicitely mentions also here:
[This name must never change in future F# compilers](https://github.com/fsharp/fslang-design/blob/main/FSharp-4.6/FS-1030-anonymous-records.md#type-identity)

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/Misc/AnonRecd.fs.il.net472.bsl (hunk 1 above)
Comment (abonie) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/Misc/AnonRecd.fs.il.net472.bsl:82]: Yup, it does not work cross-assembly :(

**Comment by abonie** [comment]
> Postponing until we get a clarification/test/.. on how this affects name stability.
> 
> The RFC explicitely mentions also here: [This name must never change in future F# compilers](https://github.com/fsharp/fslang-design/blob/main/FSharp-4.6/FS-1030-anonymous-records.md#type-identity)

I have addressed this now. It's a bit clunky - we use new modified stamps everywhere except for generating type names, where we use the old recipe - just to make sure it does not affect backward compatibility.

**Review by KevinRansom — APPROVED** [review]
Looks great.

**Pull request merged**